fix: Unity.Netcode.RuntimeTests Failing on Console

NoelStephensUnity opened the pull request:
This should fix the issues when running Unity.Netcode.RuntimeTests in a stand alone test runner build.  These changes should get all console platforms passing for this assembly.
[MTT-2288](https://jira.unity3d.com/browse/MTT-2288)

With the fixes in place [all consoles are passing Unity.Netcode.RuntimeTests](https://unity-ci.cds.internal.unity3d.com/job/11430444/dependency-graph):
![image](https://user-images.githubusercontent.com/73188597/152630040-48fc1f2d-d468-4a69-ad87-51d204f39790.png)

## Testing and Documentation
* No tests have been added.
* No documentation changes or additions were necessary.

NoelStephensUnity committed:
fix

This should fix the timing issues on consoles for the NetworkVariableTests.

NoelStephensUnity committed:
style

Just added some additional comments.

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
fix - partial

This is a "quasi-fix" where I had to add the ability to run NetworkTransform in the FixedUpdate in order to get the NetworkTransform tests all passing when running in a stand alone test runner build.

NoelStephensUnity committed:
style

Removed unused using statement.

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
fix

This includes the fixes for getting NetworkVarBufferCopyTest to run on a stand alone test runner build.

NoelStephensUnity committed:
style

white space missing.

NoelStephensUnity committed:
fix

This includes updates to the NetworkObjectOnSpawnTests in order to get it passing in stand alone test runner builds in order to get it to pass on consoles.

NoelStephensUnity committed:
fix

delta read was failing.

Cosmin-B commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -875,6 +875,23 @@ protected virtual void Update()
         }
+        // todo this is currently in update, to be able to catch any transform changes. A FixedUpdate mode could be added to be less intense, but it'd be
+        // conditional to users only making transform update changes in FixedUpdate.
+        protected virtual void Update()
```
I can understand the idea and it makes sense to me, as in what you are trying to achieve. However, I would never want to have all my hundreds of NetworkTransform objects to check each frame or even multiple times per frame whether a boolean is set to false or true.

We can improve this by having a Manager that would run the updates for them and the NetworkTransform would subscribe on Start to either FixedUpdate/Update delegates.

So, for this specific part, we can have further talks.

NoelStephensUnity replied:
Yeah...I already knew this portion of the fix would kick off a big debate... 
So, I might bring it up next week in the weekly discussion or create an earlier meeting to discuss this one particular change.   The really short version is that there isn't just a simple fix in the integration test itself for the issues that were happening on stand alone test runner builds as it has to do with the "unrealistic" timing issues that happen due to MultiInstanceHelpers, BaseMultiInstanceTest, SIPTransport, and how NetworkTransform initializes itself.
But yeah...I knew there was going to be a bunch of discussion around this fix... :)

jeffreyrainy replied:
@Cosmin-B 
> I can understand the idea and it makes sense to me, as in what you are trying to achieve. However, I would never want to have all my hundreds of NetworkTransform objects to check each frame or even multiple times per frame whether a boolean is set to false or true.

I get your point, but the cost of a boolean test is probably much lower than the call to update (or fixedUpdate) itself from the engine. Once you put the object in the engine, the cost to update it has already been incurred, no?

Cosmin-B replied:
I don't consider it lower, in a case where you have NetworkUpdateManager (which is a Unity best practice) you only pay the cost of one FixedUpdate/Update call from the engine because the rest of the calls to its subscribers would be similar to calling a C# function as you would do normally, which means that the cost of the Unity Events would not be incurred as they simply won't be there.

That is how I would be implementing these "updates" myself.

P.S.: So in the current implementation, I agree with you that the FixedUpdate/Update calls would be more expensive than the bool itself if you just leave it like that, but it means that we are adding an additional minor perf hit. Furthermore, i don't think would stop this PR to go in.

NoelStephensUnity replied:
There is a new approach that I just pushed which does not involve making adjustments to NetworkTransform.  
Still need to verify it works on consoles, but hopefully the adjustment will work and then it is just a matter of approving the changes to the test itself.

NoelStephensUnity committed:
style

removing a portion of the comment as it was a copy-paste of code used in other fixes.

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
update

Removing the auto-assignment of the NetworkTransformTests's constructor's properties that were left over from troubleshooting.

NoelStephensUnity committed:
Merge branch 'fix/console-tests-failing' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/console-tests-failing

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
update

Applied suggestion by Fatih to use deltaTime with normal Update selected and fixedDeltaTime for FixedUpdate selected.

NoelStephensUnity committed:
refactor

using the bool as opposed to direct method call for the while loops.

NoelStephensUnity committed:
refactor

moving the HasTimedOut check to the end of each while loop to assure that any possible edge case scenario where it finishes just as it would be timing out is not missed.  An example would be the condition being checked is validated but it lands right on the timeout period.  The test would have been a success, but timedOut would have been set too.

NoelStephensUnity committed:
revert

re-applying readonly to the m_TestWithClientNetworkTransform and m_TestWithHost properties as that was changed during troubleshooting.

NoelStephensUnity committed:
refactor

removing the timeout code to reduce confusion on the changes to NetworkVariableTests.

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
refactor

Came up with what I hope is a much better fix to get NetworkTransformTests passing on all consoles.  Passes on a windows stand alone test runner build...will need to run this on the consoles to verify the fix.

NoelStephensUnity committed:
style

making sure NetworkTransform was exactly as it was before, removing a LF/CR

NoelStephensUnity committed:
style

removing whitespace.

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
test refactor

This update collapses all of the time out code into the BaseMultiInstanceTest with an added method WaitForConditionOrTimeOut that leverages from an added TimeOutHelper class.

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
Merge branch 'develop' into fix/console-tests-failing

NoelStephensUnity committed:
test update and style

removing some of the Run calls and beginning to convert over some of the problematic tests to  just yield return the conditional check itself to avoid potential timing issues when checking the condition and yielding to the current test group's newly created coroutine.

NoelStephensUnity committed:
update

removing MultiInstanceHelpers.Run calls from NetworkTransform as they are not needed.

NoelStephensUnity committed:
update

removing more MultiInstanceHelpers.Run calls

NoelStephensUnity committed:
fix

The AllNetworkVariableTypes was not properly creating the NetworkManager based on the value of useHost

NoelStephensUnity committed:
style

removing a using statement for a namespace that was not being used.

NoelStephensUnity committed:
update

Reviewing the updated methods and making some minor adjustments.

NoelStephensUnity committed:
fix and update

fixed issue with WhenListContainsManyLargeValues_OverflowExceptionIsNotThrown count comparison value.

Updated removed FixedString32Test by removing  yield return that was not needed.

NoelStephensUnity committed:
style

arrg!!!!!! white space

NoelStephensUnity committed:
fix

this fixes an issue I discovered in the IndexOf test.
